compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -257,6 +263,175 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         ty
     }
 
+    /// Whether this expression constitutes a read of value of the type that
+    /// it evaluates to.
+    ///
+    /// This is used to determine if we should consider the block to diverge
+    /// if the expression evaluates to `!`, and if we should insert a `NeverToAny`
+    /// coercion for values of type `!`.
+    ///
+    /// This function generally returns `false` if the expression is a place
+    /// expression and the *parent* expression is the scrutinee of a match or
+    /// the pointee of an `&` addr-of expression, since both of those parent
+    /// expressions take a *place* and not a value.
+    ///
+    /// This function is unfortunately a bit heuristical, though it is certainly
+    /// far sounder than the prior status quo: <https://github.com/rust-lang/rust/issues/117288>.
+    pub(super) fn expr_constitutes_read(&self, expr: &'tcx hir::Expr<'tcx>) -> bool {
+        // We only care about place exprs. Anything else returns an immediate
+        // which would constitute a read. We don't care about distinguishing
+        // "syntactic" place exprs since if the base of a field projection is
+        // not a place then it would've been UB to read from it anyways since
+        // that constitutes a read.
+        if !expr.is_syntactic_place_expr() {
+            return true;
+        }
+
+        let parent_node = self.tcx.parent_hir_node(expr.hir_id);
+        match parent_node {
+            hir::Node::Expr(parent_expr) => {
+                match parent_expr.kind {
+                    // Addr-of, field projections, and LHS of assignment don't constitute reads.
+                    // Assignment does call `drop_in_place`, though, but its safety
+                    // requirements are not the same.
+                    ExprKind::AddrOf(..) | hir::ExprKind::Field(..) => false,
+                    ExprKind::Assign(lhs, _, _) => {
+                        // Only the LHS does not constitute a read
+                        expr.hir_id != lhs.hir_id
+                    }
+
+                    // If we have a subpattern that performs a read, we want to consider this
+                    // to diverge for compatibility to support something like `let x: () = *never_ptr;`.
+                    ExprKind::Match(scrutinee, arms, _) => {
+                        assert_eq!(scrutinee.hir_id, expr.hir_id);
+                        arms.iter().any(|arm| self.pat_constitutes_read(arm.pat))
```

Comment:
`any` vs `all`?

I am a bit confused by how this is supposed to work, i.e. whether we want to treat this as a 'proper soundness issue' and be conservative with what we allow, or whether it's fine to allow some code to introduce reads evne though they aren't strictly necessary :thinking: 

this allows the following, where afaik the first branch does not read from `*x` but is still treated as unreachable
```rust
#![feature(never_type)]
fn foo<'a>(x: *const !) -> ! {
    unsafe {
        match *x {
            _ => {
                let y: &'a () = &();
                let z: &'static () = y;
                return *x;
            },
            value => loop {}
        }
    }
}
```

Reply:
Well the correct way of doing this would be to implement some sort of covering algorithm with unreachable pattern arms, or at least an approximation with `_` and `!` patterns. Especially bc we only care about when the _ pattern is reading the `!` part of a complex lvalue like `(T, !)`.

Kinda think that what is implemented here is at least better than the status quo tho.

Reply:
I totally agree this is an improvement over the status quo. While I understand the difficulties in implementing this correctly, I do not understand why we err on the side of inserting *unsound* coercions instead of incorrectly errororing in very rare cases by preventing coercions which could be allowed.

Reply:
Presumably because changing the *existing* behavior (recall that we unconditionally insert these coercions today) is going to break existing code that relies on these coercions. I can try to just make a version that requires all sub patterns to be reads, and see what that breaks I guess.

Reply:
I don't think there will be basically any breakage, this requires you to match on a value of type `!` and bind it via `_`, and use incorrect assumptions we've made in the past. Seems quite unlikely, although only actual tests can tell.

Reply:
A side-effect of `all` is that `match *never_ptr {}` is considered to be a diverging read of `*never_ptr`, which I actually think is correct behavior (since it's effectively like a read of an empty enum's discriminant).